File: iOS/AgoraEntScenarios/Scenes/Show/Service/ShowServiceProtocol.swift
//
//  ShowServiceProtocol.swift
//  AgoraEntScenarios
//
//  Created by wushengtao on 2022/11/2.
//

import Foundation

enum ShowSubscribeStatus {
    case created     //订阅到对象创建
    case deleted     //订阅到对象删除
    case updated     //订阅到对象更新
}

protocol ShowServiceProtocol: NSObjectProtocol {
    /// 获取房间列表
    /// - Parameters:
    ///   - page: 分页索引，从0开始(由于SyncManager无法进行分页，这个属性暂时无效)
    ///   - completion: 完成回调   (错误信息， 房间列表)
    func getRoomList(page: Int,
                     completion: @escaping (NSError?, [ShowRoomListModel]?) -> Void)
    
    /// 创建房间
    /// - Parameters:
    ///   - room: 房间对象信息
    ///   - completion: 完成回调   (错误信息)
    
    
    /// 创建房间
    /// - Parameters:
    ///   - roomId: 房间Id 
    ///   - roomName: 房间名
    ///   - completion: 完成回调
    func createRoom(roomId: String,
                    roomName: String,
                    completion: @escaping (NSError?, ShowRoomDetailModel?) -> Void)
    
    /// 加入房间
    /// - Parameters:
    ///   - room: 房间对象信息
    ///   - completion: 完成回调   (错误信息， 房间信息)
    func joinRoom(room: ShowRoomListModel,
                  completion: @escaping (NSError?, ShowRoomDetailModel?) -> Void)
    
    
    /// 离开房间
    /// - Parameters:
    ///   - roomId: 房间Id
    ///   - completion: 完成回调
    func leaveRoom(roomId: String, 
                   completion: @escaping (NSError?) -> Void)
    
    /// 获取当前房间所有用户
    /// - Parameters:
    ///   - roomId: 房间Id
    ///   - completion: 完成回调   (错误信息， 用户列表)
    func getAllUserList(roomId: String, 
                        completion: @escaping (NSError?, [ShowUser]?) -> Void)
    
    /// 发送聊天消息
    /// - Parameters:
    ///   - roomId: 房间Id
    ///   - message: 消息内容
    ///   - completion: 完成回调
    func sendChatMessage(roomId: String,
                         message: ShowMessage,
                         completion: ((NSError?) -> Void)?)
    
    /// 获取上麦申请列表
    /// - Parameters:
    ///   - roomId: 房间Id
    ///   - completion: 完成回调
    func getAllMicSeatApplyList(roomId: String, 
                                completion: @escaping (NSError?, [ShowMicSeatApply]?) -> Void)

    
    /// 观众申请连麦
    /// - Parameters:
    ///   - roomId: 房间Id
    ///   - completion: 完成回调
    func createMicSeatApply(roomId: String,
                            completion: @escaping (NSError?) -> Void)
    
    /// 观众取消连麦申请
    /// - Parameters:
    ///   - roomId: 房间Id
    ///   - completion: 完成回调
    func cancelMicSeatApply(roomId: String,
                            completion: @escaping (NSError?) -> Void)
    
    /// 主播接受连麦申请
    /// - Parameters:
    ///   - roomId: 房间Id
    ///   - userId: 用户Id
    ///   - completion: 完成回调
    func acceptMicSeatApply(roomId: String,
                            userId: String,
                            completion: @escaping (NSError?) -> Void)
    
    
    
    
    /// 主播创建连麦邀请
    /// - Parameters:
    ///   - roomId: 房间ID
    ///   - userId: 邀请用户id
    ///   - completion: 完成回调
    func createMicSeatInvitation(roomId: String,
                                 userId: String,
                                 completion: @escaping (NSError?) -> Void)
    
    
    /// 观众同意连麦
    /// - Parameters:
    ///   - roomId: 房间ID
    ///   - invitationId: 邀请id
    ///   - completion: 完成回调
    func acceptMicSeatInvitation(roomId: String,
                                 invitationId: String,
                                 completion: @escaping (NSError?) -> Void)
    
    /// 观众拒绝连麦
    /// - Parameters:
    ///   - roomId: 房间ID
    ///   - invitationId: 邀请id
    ///   - completion: 完成回调
    func rejectMicSeatInvitation(roomId: String,
                                 invitationId: String,
                                 completion: @escaping (NSError?) -> Void)
    
    
    
    
    /// 获取可PK对象列表(目前等价getRoomList)
    /// - Parameter completion: 完成回调
    func getAllPKUserList(completion: @escaping (NSError?, [ShowPKUserInfo]?) -> Void)
    
    /// 创建PK邀请
    /// - Parameters:
    ///   - roomId: 房间id
    ///   - pkRoomId: 需要pk的房间id
    ///   - completion: 完成回调
    func createPKInvitation(roomId: String,
                            pkRoomId: String,
                            completion: @escaping (NSError?) -> Void)
    
    /// 同意PK
    /// - Parameters:
    ///   - roomId: 房间id
    ///   - invitationId: 邀请id
    ///   - completion: 完成回调
    func acceptPKInvitation(roomId: String,
                            invitationId: String,
                            completion: @escaping (NSError?) -> Void)
    
    /// 拒绝PK
    /// - Parameters:
    ///   - roomId: 房间id
    ///   - invitationId: 邀请id
    ///   - completion: 完成回调
    func rejectPKInvitation(roomId: String,
                            invitationId: String,
                            completion: @escaping (NSError?) -> Void)
    
    
    
    
    /// 获取当前互动信息
    /// - Parameters:
    ///   - roomId: 房间id
    ///   - completion: 完成回调
    func getInterationInfo(roomId: String,
                           completion: @escaping (NSError?, ShowInteractionInfo?) -> Void)
    
    /// 停止互动
    /// - Parameters:
    ///   - roomId: 房间id
    ///   - completion: 完成回调
    func stopInteraction(roomId: String,
                         completion: @escaping (NSError?) -> Void)
    
    
    
    
    /// 静音设置
    /// - Parameters:
    ///   - mute: <#mute description#>
    ///   - userId: <#userId description#>
    ///   - completion: <#completion description#>
    func muteAudio(roomId: String,
                   mute:Bool,
                   completion: @escaping (NSError?) -> Void)

    
    func getCurrentNtpTs(roomId: String) -> UInt64
    
    func unsubscribeEvent(roomId: String, delegate: ShowSubscribeServiceProtocol)
    
    func subscribeEvent(roomId: String, delegate: ShowSubscribeServiceProtocol)
}

File: Android/scenes/show/src/main/java/io/agora/scene/show/service/ShowServiceProtocol.kt
package io.agora.scene.show.service

import io.agora.scene.base.component.AgoraApplication

/*
 * Service module
 * Introduction: This module is responsible for interaction between frontend business modules and business server 
 * (including room list + room business data synchronization, etc.)
 * Implementation principle: The business server of this scene is wrapped with a rethinkDB backend service for data storage.
 * It can be considered as a DB that can be freely written by the app side. Room list data and room business data 
 * are constructed on the app and stored in this DB.
 * When data in DB is added/deleted/modified, all clients will be notified to achieve business data synchronization
 * TODO Note⚠️: The backend service of this scene is for demo only and cannot be used commercially. 
 * If you need to go live, you must deploy your own backend service or cloud storage server 
 * (such as leancloud, easemob, etc.) and reimplement this module!!!!!!!!!
 */
interface ShowServiceProtocol {

    companion object {
        // Room survival time in milliseconds
        var ROOM_AVAILABLE_DURATION: Long = 1200 * 1000
        // PK session duration in milliseconds
        var PK_AVAILABLE_DURATION: Long = 120 * 1000

        private var instance : ShowServiceProtocol? = null
            get() {
                if (field == null) {
                    field = ShowServiceImpl(AgoraApplication.the())
                }
                return field
            }

        @Synchronized
        fun get(): ShowServiceProtocol = instance!!

        @Synchronized
        fun destroy() {
            (instance as? ShowServiceImpl)?.destroy()
            instance = null
        }
    }

    // Get room list
    fun getRoomList(
        success: (List<ShowRoomDetailModel>) -> Unit,
        error: ((Exception) -> Unit)? = null
    )

    // Create room
    fun createRoom(
        roomId: String,
        roomName: String,
        success: (ShowRoomDetailModel) -> Unit,
        error: ((Exception) -> Unit)? = null
    )

    // Join room
    fun joinRoom(
        roomId: String,
        success: () -> Unit,
        error: ((Exception) -> Unit)? = null
    )

    // Leave room
    fun leaveRoom(roomId: String)

    // Subscribe to update/delete events of current joined room
    fun subscribeCurrRoomEvent(roomId: String, onUpdate: (status: ShowSubscribeStatus, roomInfo: ShowRoomDetailModel?) -> Unit)

    // Get all users in current room
    fun getAllUserList(roomId: String, success: (List<ShowUser>) -> Unit, error: ((Exception) -> Unit)? = null)

    // Monitor user changes
    fun subscribeUser(roomId: String, onUserChange: (ShowSubscribeStatus, ShowUser?) -> Unit)

    // Send chat message
    fun sendChatMessage(
        roomId: String,
        message: String,
        success: ((ShowMessage) -> Unit)? = null,
        error: ((Exception) -> Unit)? = null
    )

    // Subscribe to chat message changes
    fun subscribeMessage(
        roomId: String,
        onMessageChange: (ShowSubscribeStatus, ShowMessage) -> Unit
    )

    // Get linking application list
    fun getAllMicSeatApplyList(
        roomId: String,
        success: (List<ShowMicSeatApply>) -> Unit,
        error: ((Exception) -> Unit)? = null
    )

    // Host subscribes to linking application changes
    fun subscribeMicSeatApply(roomId: String, onMicSeatChange: (ShowSubscribeStatus, List<ShowMicSeatApply>) -> Unit)

    // Audience applies for linking
    fun createMicSeatApply(roomId: String, success: ((ShowMicSeatApply) -> Unit)? = null, error: ((Exception) -> Unit)? = null)

    // Audience cancels linking application
    fun cancelMicSeatApply(roomId: String, success: (() -> Unit)? = null, error: ((Exception) -> Unit)? = null)

    // Host accepts linking application
    fun acceptMicSeatApply(
        roomId: String,
        userId: String,
        success: (() -> Unit)? = null,
        error: ((Exception) -> Unit)? = null
    )

    // Audience subscribes to linking invitation
    fun subscribeMicSeatInvitation(
        roomId: String,
        onMicSeatInvitationChange: (ShowSubscribeStatus, ShowMicSeatInvitation?) -> Unit
    )

    // Host creates linking invitation
    fun createMicSeatInvitation(
        roomId: String,
        userId: String,
        success: ((ShowMicSeatInvitation) -> Unit)? = null,
        error: ((Exception) -> Unit)? = null
    )

    // Audience accepts linking
    fun acceptMicSeatInvitation(
        roomId: String,
        invitationId: String,
        success: (() -> Unit)? = null,
        error: ((Exception) -> Unit)? = null
    )

    // Audience rejects linking
    fun rejectMicSeatInvitation(
        roomId: String,
        invitationId: String,
        success: (() -> Unit)? = null,
        error: ((Exception) -> Unit)? = null
    )

    // Get available PK target list
    fun getAllPKUserList(
        roomId: String,
        success: ((List<ShowPKUser>) -> Unit),
        error: ((Exception) -> Unit)? = null
    )

    // Audience subscribes to PK invitation
    fun subscribePKInvitationChanged(roomId: String, onPKInvitationChanged: (ShowSubscribeStatus, ShowPKInvitation?) -> Unit)

    // Create PK invitation
    fun createPKInvitation(
        roomId: String,
        pkRoomId: String,
        success: ((ShowPKInvitation) -> Unit)? = null,
        error: ((Exception) -> Unit)? = null
    )

    // Accept PK
    fun acceptPKInvitation(
        roomId: String,
        invitationId: String,
        success: (() -> Unit)? = null,
        error: ((Exception) -> Unit)? = null
    )

    // Reject PK
    fun rejectPKInvitation(
        roomId: String,
        invitationId: String,
        success: (() -> Unit)? = null,
        error: ((Exception) -> Unit)? = null
    )

    // Get current interaction information
    fun getInteractionInfo(
        roomId: String,
        success: ((ShowInteractionInfo?) -> Unit)? = null,
        error: ((Exception) -> Unit)? = null
    )

    // Subscribe to interaction invitation
    fun subscribeInteractionChanged(roomId: String, onInteractionChanged: (ShowSubscribeStatus, ShowInteractionInfo?) -> Unit)

    // Stop interaction
    fun stopInteraction(
        roomId: String,
        success: (() -> Unit)? = null,
        error: ((Exception) -> Unit)? = null
    )

    // Mute settings
    fun muteAudio(
        roomId: String,
        mute: Boolean,
        success: (() -> Unit)? = null,
        error: ((Exception) -> Unit)? = null
    )

    // Subscribe to reconnection events
    fun subscribeReConnectEvent(roomId: String, onReconnect: () -> Unit)

    // Start robot
    fun startCloudPlayer()
}